companies/views.py

```diff
@@ -1,86 +1,144 @@
-from rest_framework_simplejwt.authentication import JWTAuthentication
-
-JWT_authenticator = JWTAuthentication()
-
-
-class CompaniesListCreateView(generics.ListCreateAPIView):
-    queryset = Company.objects.all()
-    serializer_class = CompaniesSerializer
-    # permission_classes = (IsAuthenticated,)
-    filter_backends = [DjangoFilterBackend]
-    filterset_class = CompanyFilter
-
+from .serializers import CompaniesSerializer, SubscriptionSerializer
+from .managers import ArticlesManager as am, LIMIT
+from forum import errors as er
+from pydantic import ValidationError as PydanticValidationError
 
-class CompaniesRetrieveUpdateView(generics.RetrieveUpdateDestroyAPIView):
-    queryset = Company.objects.all()
-    serializer_class = CompaniesSerializer
-    # permission_classes = (IsAuthenticated,)
-    filter_backends = [DjangoFilterBackend]
-    filterset_class = CompanyFilter
 
-
-class SubscriptionCreateAPIView(APIView):
+class CompaniesListCreateView(APIView):
     permission_classes = (IsAuthenticated,)
 
-    def post(self, request, *args, **kwargs):
-
-        auth_response = JWT_authenticator.authenticate(request)
-        if auth_response is None:
-            return Response({'error': "Invalid token"}, status=status.HTTP_401_UNAUTHORIZED)
-        user_id = auth_response[1].get('user_id')
-        request.data['investor'] = user_id
-        serializer = SubscriptionSerializer(data=request.data)
+    def post(self, request):
+        serializer = CompaniesSerializer(data=request.data)
         if serializer.is_valid():
-            subs = Subscription.objects.filter(investor=user_id, company=request.data.get('company'))
-            if not subs:
-                subscription = serializer.save()
-            else:
-                subscription = subs.first()
-            company = subscription.company
-            brand = company.brand
-            subscription_id = subscription.subscription_id
-            message = f"You're subscribed to {brand}, subscription id: {subscription_id}"
-            return Response({'message': message}, status=status.HTTP_201_CREATED)
-        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
+            serializer.save()
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        else:
+            return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
 
-
-class UnsubscribeAPIView(APIView):
+class CompanyRetrieveView(APIView):
+    permission_classes = (IsAuthenticated,)
+    def get(self, request, pk=None):
+        if pk:
+           try:
+               company = Company.get_company(company_id=pk)
+               return Response(company.get_info(), status=status.HTTP_200_OK)
+           except Company.DoesNotExist:
+               return Response({'error': 'Company not found'}, status=status.HTTP_404_NOT_FOUND) 
+        else: 
+            return Response({'error': er.NO_CREDENTIALS.msg}, status=er.NO_CREDENTIALS.status)
```

Comment:
The error message and status code seem to imply missing credentials, but the actual issue is a missing company ID (pk). Consider using a more appropriate error message, such as "Company ID required" with a `status.HTTP_400_BAD_REQUEST`

Reply:
Agree, fixed.